### PR TITLE
Fix SAM template indexing crashes when the required properties are null.

### PR DESCRIPTION
**Pull request by zhangzhx**
Fix the bug that when the required properties, eg Runtime in Lambda Handler, don't exist in the sam template file, the indexing errors out.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Motivation and Context

The exception stack trace for the error.
```
ERROR: can't parse argument number: 
java.lang.IllegalArgumentException: can't parse argument number: 
	at java.text.MessageFormat.makeFormat(MessageFormat.java:1429)
	at java.text.MessageFormat.applyPattern(MessageFormat.java:479)
	at java.text.MessageFormat.<init>(MessageFormat.java:362)
	at java.text.MessageFormat.format(MessageFormat.java:840)
	at software.aws.toolkits.resources.Localization.message(Localization.kt:19)
	at software.aws.toolkits.jetbrains.services.cloudformation.yaml.YamlCloudFormationTemplate$YamlResource.getScalarProperty(YamlCloudFormationTemplate.kt:57)
	at software.aws.toolkits.jetbrains.services.cloudformation.SamFunction.getScalarProperty(Resources.kt)
	at software.aws.toolkits.jetbrains.services.cloudformation.IndexedResource.<init>(IndexedResources.kt:14)
	at software.aws.toolkits.jetbrains.services.cloudformation.IndexedFunction.<init>(IndexedResources.kt:39)
	...
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
Caused by: java.lang.NumberFormatException: For input string: ""
	at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
	at java.lang.Integer.parseInt(Integer.java:592)
	at java.lang.Integer.parseInt(Integer.java:615)
	at java.text.MessageFormat.makeFormat(MessageFormat.java:1427)
	... 120 more
Error while indexing /src/template.yaml
```

## Checklist

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `gradlew check` succeeds
- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/IndexedResources.kt [hunk 1]
```diff
@@ -10,7 +10,13 @@ package software.aws.toolkits.jetbrains.services.cloudformation
 open class IndexedResource internal constructor(val indexedProperties: Map<String, String>) {
 
     protected constructor(resource: Resource, indexProperties: List<String>)
-            : this(indexProperties.map { it to resource.getScalarProperty(it) }.toMap())
+            : this(indexProperties.map {
+        it to try {
+            resource.getScalarProperty(it)
+        } catch (e: Exception) {
+            ""
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/IndexedResources.kt]: I'd rather have `null` than empty string - at least that's something that the compiler forces me to handle.

Comment (zhangzhx) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/IndexedResources.kt]: Changed to `null` and also refactored the persistence code to the `IndexedResource` class

Comment (kiiadi) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/IndexedResources.kt]: Did you push the code? Looks like it's still returning empty string...

Comment (zhangzhx) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/IndexedResources.kt]: Yes, I did push two separate commits. Any more concerns?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/IndexedResources.kt [hunk 2]
```diff
@@ -3,38 +3,72 @@
...
 /**
  * Immutable data class for indexing [Resource]. Use [from] to create an instance so that it always
  * returns a concrete [IndexedResource] such as [IndexedFunction] if applicable.
  */
-open class IndexedResource internal constructor(val indexedProperties: Map<String, String>) {
+open class IndexedResource protected constructor(val type: String, val indexedProperties: Map<String, String>) {
 
     protected constructor(resource: Resource, indexProperties: List<String>)
-            : this(indexProperties.map { it to resource.getScalarProperty(it) }.toMap())
+        : this(resource.type() ?: throw Exception(message("cloudformation.template_index.missing_type")),
```
[the remark this answers is not in the document]
Comment (zhangzhx) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudformation/IndexedResources.kt]: Done.

**Pull request merged**